packages/ember-metal/lib/meta_listeners.js

```diff
@@ -0,0 +1,154 @@
+/*
+ When we render a rich template hierarchy, the set of events that
+ *might* happen tends to be much larger than the set of events that
+ actually happen. This implies that we should make listener creation &
+ destruction cheap, even at the cost of making event dispatch more
+ expensive.
+
+ Thus we store a new listener with a single push and no new
+ allocations, without even bothering to do deduplication -- we can
+ save that for dispatch time, if an event actually happens.
+ */
+
+/* listener flags */
+export var ONCE = 1;
+export var SUSPENDED = 2;
+
+export var protoMethods = {
+
+  addToListeners(eventName, target, method, flags) {
+    if (!this._listeners) {
+      this._listeners = [];
+    }
+    this._listeners.push(eventName, target, method, flags);
+  },
+
+  _finalizeListeners() {
```

Comment:
Well something worth noting, deletion of instance observers does happen often. One of the biggest perf issue we have (after initial render) is actually transitioning and destroying the stuff from the previous route.